### PR TITLE
cli: refactor runQuit().

### PR DESCRIPTION
**Pull request by knz**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7640)

### CODE REVIEW

**Comment by tamird** [comment]
This is more readable?

**Comment by knz** [comment]
7 less lines of code, reduced code duplication, the logic flow of the shutdown operation in runQuit() is now easier to grasp, I think this PR matches a couple of objective criteria, doesn't it?

**Comment by tamird** [comment]
+40, -42, so the difference is two lines of code. I find this flow harder to grasp, not easier.

Perhaps I'm the wrong reviewer for this.

**Comment by knz** [comment]
There are +5 lines of comments... Look you asked me yesterday to provide an alternative, and so I did. I'm really trying to help here.

**Comment by tamird** [comment]
I'm not saying you aren't - I'm just finding this harder to read, but that might be because I wrote the original. That's why I suggested that I'm not the right reviewer.

File: cli/start.go [hunk 1]
```diff
@@ -480,6 +480,39 @@ completed, the server exits.
 	RunE:         runQuit,
 }
 
+// waitForShutdown waits until the gRPC streams is closed by the
+// remote end.
+func waitForShutdown(stream serverpb.Admin_DrainClient) error {
```
Comment (tschottdorf) [inline cli/start.go]: I would inline this.

**Comment by tschottdorf** [comment]
LGTM after you inline this. I think the main function benefits from outsourcing `doShutdown`.

**Comment by tschottdorf** [comment]
Hold on a sec, some more comments.

File: cli/start.go [hunk 2]
```diff
@@ -480,6 +480,39 @@ completed, the server exits.
...
+	}
+	_, err = stream.Recv()
+	if err == nil {
+		err = waitForShutdown(stream)
+		if err == nil {
+			fmt.Println("ok")
+		}
+		return false, err
+	}
+	return true, err
```
Comment (tschottdorf) [inline cli/start.go]: Don't return `err` here. If this method returns an error, that's it. If it doesn't return an error, the caller gets to check the boolean (and maybe retry).

File: cli/start.go [hunk 3]
```diff
@@ -495,53 +528,18 @@ func runQuit(_ *cobra.Command, _ []string) error {
...
-			err = waitForShutdown(stream)
-			if err == nil {
-				fmt.Println("ok")
-			}
-			return err
-		}
-		fmt.Printf("graceful shutdown failed, proceeding with hard shutdown: %v\n", err)
-	}
+	retry, err := doShutdown(c, ctx, onModes)
+	if retry {
```
Comment (tschottdorf) [inline cli/start.go]: check `err` before `retry` (see above comment).

File: cli/start.go (hunk 2 above)
Comment (knz) [inline cli/start.go]: I don't fully agree with this. The point of the current structure is that the caller gets an err object to print the intermediate error message "soft shutdown failed, trying hard shutdown instead..." whjch embeds the intermediate error.

File: cli/start.go (hunk 1 above)
Comment (knz) [inline cli/start.go]: Done

**Comment by knz** [comment]
PTAL

**Comment by tschottdorf** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7640#-:-KMB-BEDeLfFQsMCN8tK:1417088721)*

**Pull request merged**